### PR TITLE
CM-778 - Add multicaller contract with tests

### PR DESCRIPTION
**Pull request by alejoloaiza**
The GuardedMulticaller contract will be deployable from Hub and therefor it needs to be included in the `contracts` repo. I've attached the following to this PR:
- Contract code
- Tests
- Threat model
- Audit report.

### CODE REVIEW

**Comment by drinkcoffee** [comment]
Could you fix the slither and solhint errors that are related to this PR.

File: contracts/mocks/MockFunctions.sol [hunk 1]
```diff
@@ -0,0 +1,16 @@
+// SPDX-License-Identifier: Unlicense
+pragma solidity ^0.8.19;
+
+contract MockFunctions {
```
Comment (drinkcoffee) [inline contracts/mocks/MockFunctions.sol:4]: Please add a contract level comment indicating that this file is part of the test code for GuardedMulticaller.

**Comment by drinkcoffee** [comment]
Add a README.md file in contracts/multicall to describe the contents of the directory. For example, see: https://github.com/immutable/contracts/tree/main/contracts/token/erc1155

In the README, the threat model and the external audit should be referenced. I expect that the version of code that was analysed for the threat model is not available in the contracts repo. As such, I would just put --- . For the external audit, you will need the git hash of the code that matches what was audited. I expect that is the code that was moved across, prior to the fixes for solhint and slither.

File: audits/multicall/202309-threat-model-multicaller/202309-threat-model-multicaller.md [hunk 2]
```diff
@@ -0,0 +1,78 @@
+## Background
```
Comment (drinkcoffee) [inline audits/multicall/202309-threat-model-multicaller/202309-threat-model-multicaller.md:1]: This .md file should be in ./audits/multicall/

Comment (alejoloaiza) [reply audits/multicall/202309-threat-model-multicaller/202309-threat-model-multicaller.md]: done

File: contracts/mocks/MockFunctions.sol (hunk 1 above)
Comment (alejoloaiza) [reply contracts/mocks/MockFunctions.sol:5]: done

**Review by drinkcoffee — APPROVED** [review]

**Review by luads — APPROVED** [review]

**Review by rytimx — APPROVED** [review]
LGTM, thanks.

**Pull request merged**